netlink/conntracker.go

```diff
@@ -0,0 +1,266 @@
+		state:         make(map[connKey]*IPTranslation),
+	}
+
+	// seed the state
+	sessions, err := nfct.Dump(ct.Ct, ct.CtIPv4)
+	if err != nil {
+		return nil, err
+	}
+	ctr.registerUnlocked(sessions)
+
+	sessions, err = nfct.Dump(ct.Ct, ct.CtIPv6)
+	if err != nil {
+		// this is not fatal because we've already seeded with IPv4
+		log.Errorf("Failed to dump IPv6")
+	}
+	ctr.registerUnlocked(sessions)
+
+	go ctr.emitStats()
+	go ctr.compact()
+
+	nfct.Register(context.Background(), ct.Ct, ct.NetlinkCtNew|ct.NetlinkCtExpectedNew|ct.NetlinkCtUpdate, ctr.register)
+	nfct.Register(context.Background(), ct.Ct, ct.NetlinkCtDestroy, ctr.unregister)
+
+	log.Infof("initialized conntrack")
+
+	return ctr, nil
+}
+
+func (ctr *realConntracker) GetConntrackEntryForConn(ip string, port uint16) *IPTranslation {
+	then := time.Now().UnixNano()
+
+	ctr.l.Lock()
+	defer ctr.l.Unlock()
+
+	result := ctr.state[connKey{ip, port}]
+
+	now := time.Now().UnixNano()
+	atomic.AddInt64(&ctr.stats.gets, 1)
+	atomic.AddInt64(&ctr.stats.getTimeTotal, now-then)
+	return result
+}
+
+func (ctr *realConntracker) Close() {
+	ctr.statsTicker.Stop()
+	ctr.compactTicker.Stop()
+	ctr.nfct.Close()
+}
+
+func (ctr *realConntracker) registerUnlocked(sessions []ct.Conn) {
+	for _, c := range sessions {
+		if isNAT(c) {
+			ctr.state[formatKey(c)] = formatIPTranslation(c)
+		}
+	}
+}
+
+func (ctr *realConntracker) register(c ct.Conn) int {
+	// don't both storing if the connection is not NAT
+	if !isNAT(c) {
+		return 0
+	}
+
+	now := time.Now().UnixNano()
+	ctr.l.Lock()
+	defer ctr.l.Unlock()
+
+	ctr.state[formatKey(c)] = formatIPTranslation(c)
+
+	then := time.Now().UnixNano()
+	atomic.AddInt64(&ctr.stats.registers, 1)
+	atomic.AddInt64(&ctr.stats.registersTotalTime, then-now)
+
+	return 0
+}
+
+func (ctr *realConntracker) unregister(c ct.Conn) int {
+	if !isNAT(c) {
+		return 0
+	}
+
+	ctr.l.Lock()
+	defer ctr.l.Unlock()
+
+	delete(ctr.state, formatKey(c))
+
+	return 0
+}
+
+func (ctr *realConntracker) compact() {
+	for range ctr.compactTicker.C {
+
+		ctr.l.Lock()
+
+		copied := make(map[connKey]*IPTranslation, len(ctr.state))
+		for k, v := range ctr.state {
+			copied[k] = v
+		}
+		ctr.state = copied
+
+		ctr.l.Unlock()
+
+	}
+}
+
+func (ctr *realConntracker) emitStats() {
+	for range ctr.statsTicker.C {
+
+		ctr.l.Lock()
+		size := len(ctr.state)
+		ctr.l.Unlock()
+
+		log.Debugf("state size %d", size)
+		if ctr.stats.gets != 0 {
+			log.Debugf("total gets: %d, ns/get: %f", ctr.stats.gets, float64(ctr.stats.getTimeTotal)/float64(ctr.stats.gets))
+		}
+		if ctr.stats.registers != 0 {
+			log.Debugf("total registers: %d, ns/register: %f", ctr.stats.registers, float64(ctr.stats.registersTotalTime)/float64(ctr.stats.registers))
+		}
+		atomic.StoreInt64(&ctr.stats.gets, 0)
+		atomic.StoreInt64(&ctr.stats.getTimeTotal, 0)
+		atomic.StoreInt64(&ctr.stats.registers, 0)
+		atomic.StoreInt64(&ctr.stats.registersTotalTime, 0)
+	}
+}
+
+func isNAT(c ct.Conn) bool {
```

Comment:
Looks like we're calling these conversions multiple times `c.OrigSrcIP()`. Does it make sense to have this return more than just a `bool`?